<?xml version="1.0"?>
<!--
data_recorder.launch: data recorder for calibration
Author: Ravi Joshi
Date: 2018/02/16
-->
<launch>
  <!-- make sure it matches with argument 'ee_topic' provided to 'data_collector' below -->
  <arg name="limb" default="right"/>
  <!-- wait time to stablize arm before capturing point cloud -->
  <arg name="wait_time" default="2"/>
  <!-- baxter arm pre-defined trajectory -->
  <arg name="trajectory" default="$(find test_package)/files/viapoints.csv"/>
  <node name="baxter_controller" pkg="test_package" type="baxter_controller.py" output="screen">
    <param name="limb" value="$(arg limb)"/>
    <param name="wait_time" value="$(arg wait_time)"/>
    <param name="trajectory" value="$(arg trajectory)"/>
  </node>

  <!-- minimum and maximum HSV values for sphere segmentation -->
  <arg name="min_hsv" default="[60, 0.6, 0.3]"/>
  <arg name="max_hsv" default="[64, 0.8, 1.0]"/>

  <!--
  ###################################################################
  # FOLLOWING CONFIGURATIONS ARE USED IN 'SAC_RANSAC' USING         #
  # 'SACMODEL_NORMAL_SPHERE'. SEE BELOW URL FOR MORE INFO-          #
  # http://docs.pointclouds.org/1.7.0/group__sample__consensus.html #
  ###################################################################
  -->
  <!-- radius of sphere (in meter) -->
  <arg name="radius" default="0.034"/>
  <!-- length of stick to hold the sphere (in meter) -->
  <arg name="offset" default="0.051"/>
  <!-- number of 'k' nearest neighbors to use for feature estimation -->
  <arg name="k_neighbors" default="10"/>
  <!-- normal angular distance weight -->
  <arg name="weight" default="0.05"/>
  <!-- maximum number of iterations before giving up -->
  <arg name="max_itr" default="1000"/>
  <!-- distance to the model threshold -->
  <arg name="d_thresh" default="0.005"/>
  <!-- probability of choosing at least one sample free from outliers -->
  <arg name="prob" default="0.99999"/>
  <!-- tolerance in radius (in meters) -->
  <arg name="tolerance" default="0.02"/>
  <!-- angle epsilon (delta) threshold (in degree) -->
  <arg name="epsilon" default="15"/>
  <!-- camera file for point cloud visualization -->
  <arg name="cam_file" default="$(find test_package)/files/camera_parameters.cam"/>
  <!-- rostopic for receving baxter ee pose. make sure it matches with argument 'limb'
  provided to 'baxter_controller' above  -->
  <arg name="ee_topic" default="/robot/limb/right/endpoint_state"/>
  <!-- rostopic for receving point cloud -->
  <arg name="pc_topic" default="/kinect2/sd/points"/>
  <!-- directory for saving collected data -->
  <arg name="data_dir" default="$(find test_package)/files/"/>
  <node name="data_collector" pkg="test_package" type="data_collector" output="screen">
    <param name="min_hsv" value="$(arg min_hsv)"/>
    <param name="max_hsv" value="$(arg max_hsv)"/>
    <param name="radius" value="$(arg radius)"/>
    <param name="offset" value="$(arg offset)"/>
    <param name="k_neighbors" value="$(arg k_neighbors)"/>
    <param name="weight" value="$(arg weight)"/>
    <param name="max_itr" value="$(arg max_itr)"/>
    <param name="d_thresh" value="$(arg d_thresh)"/>
    <param name="prob" value="$(arg prob)"/>
    <param name="tolerance" value="$(arg tolerance)"/>
    <param name="epsilon" value="$(arg epsilon)"/>
    <param name="cam_file" value="$(arg cam_file)"/>
    <param name="ee_topic" value="$(arg ee_topic)"/>
    <param name="pc_topic" value="$(arg pc_topic)"/>
    <param name="data_dir" value="$(arg data_dir)"/>
  </node>
</launch>
